src/coredump/coredump-ratelimit.c

```diff
@@ -0,0 +1,170 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <sys/statvfs.h>
+#include "dirent-util.h"
+#include "fd-util.h"
+#include "fs-util.h"
+#include "string-util.h"
+#include "time-util.h"
+#include "user-util.h"
+#include "util.h"
+#include "coredump-ratelimit.h"
+#include "sd-id128.h"
+
+#define MAX_COMM_LENGTH 16
+
+static int comm_from_file_name(const char *file_name, char *comm) {
+        const char *p;
+        int i = 0;
+
+        /* check if the file is a core dump file */
+        p = startswith(file_name, "core.");
+        if (!p) {
+                /* check if the file is an inprogress core dump file */
+                p = startswith(file_name, ".#core.");
+                if(!p)
+                        return -EINVAL;
+        }
+
+        /* Retrieve the process name (COMM) */
+        while (*p != '.' && i < MAX_COMM_LENGTH){
+                comm[i] = *p;
+                p++;
+                i++;
+        }
+        comm[i] = '\0';
+        return 0;
+}
+
+static int bootid_from_file_name(const char *file_name, char *bootid) {
+        const char *p;
+        int i = 0;
+
+        /* check if the file is a core dump file */
+        p = startswith(file_name, "core.");
+        if (!p) {
+                /* check if the file is an inprogress core dump file */
+                p = startswith(file_name, ".#core.");
+                if(!p)
+                        return -EINVAL;
+        }
+
+        /* Skip the comm field */
+        p = strchr(p, '.');
+        if (!p)
+                return -EINVAL;
+        p++;
+
+        /* Skip the uid field */
+        p = strchr(p, '.');
+        if (!p)
+                return -EINVAL;
+        p++;
+
+        /* Retrieve the Bootid */
+        while (*p != '.' && i < SD_ID128_STRING_MAX - 1){
+                bootid[i] = *p;
+                p++;
+                i++;
+        }
+        bootid[i] = '\0';
+        return 0;
+}
+
+int coredump_ratelimit(const char* process_name, usec_t interval, unsigned burst) {
+
+        _cleanup_closedir_ DIR *d = NULL;
+        struct dirent *de;
+        unsigned int count = 0;
+        usec_t curr_ts;
+        sd_id128_t boot = {};
+        char boot_id[SD_ID128_STRING_MAX];
+        int r;
+
+        /* If the process name is not given, then return -1. No rate limit happens.*/
+        if(process_name == NULL)
+                return -1;
+
+        /* either if burst or interval is zero then ratelimit is disabled */
+        if(burst == 0 || interval == 0)
+                return 0;  /* No ratelimit */
+
+        /* Whenever a daemon crashes, we compare the number of core dump files present in the
+         * /var/lib/systemd/coredump belonging to that daemon and generated within the
+         * rate limit interval with the rate limit. If the limit is reached, the core dump
+         * will not be processed further for generation.
+         */
+
+        d = opendir("/var/lib/systemd/coredump");
+
```

Comment:
remove this empty line. We generally try to keep function calls and their immediate error handling together, i.e. it's fine to place empty lines before the function call that can fail, and after the error handling, but please keep the function call and the error handling reasonably together

Reply:
Addressed the comment. Please check once.